### PR TITLE
Handle push rejection message in Merge & Web Editor

### LINKED ISSUE #10322: 500 error if PR merge is rejected by pre-receive hook (exit 1)
- Gitea version (or commit ref): 1.11.1 (docker image)
- Git version: not relevant
- Operating system: not relevant
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist:

## Description
I tried to test *pre-receive* git hook (by editing via gitea web interface) and setup simple rejection
```javascript
#!/bin/sh
echo "Forced rejection" >&2
exit 1
```
every direct (again via web interface) file modification is rejected as expected with the error message above. However if I try to merge a pull request I get 500 error page instead.

### PR DESCRIPTION
**Pull request by zeripath**
Merge push rejection due to pre-receive hook failure should not necessarily result in 500.

Fix: #10322

### CODE REVIEW

File: models/error.go [hunk 1]
```diff
@@ -1371,6 +1372,48 @@ func (err ErrMergePushOutOfDate) Error() string {
 	return fmt.Sprintf("Merge PushOutOfDate Error: %v: %s\n%s", err.Err, err.StdErr, err.StdOut)
 }
 
+// ErrPushRejected represents an error if merging fails due to unrelated histories
```
Comment (guillep2k) [inline models/error.go]: However it's not always due to unrelated histories? (e.g. hook is failing)

File: models/error.go [hunk 2]
```diff
@@ -1371,6 +1372,48 @@ func (err ErrMergePushOutOfDate) Error() string {
...
+// Message returns the cleaned message from Error
+func (err ErrPushRejected) Message() string {
+	messageBuilder := &strings.Builder{}
+	i := 0
+	for {
+		if len(err.StdErr) <= i+8 {
+			break
+		}
+		if err.StdErr[i:i+8] != "remote: " {
+			break
```
Comment (guillep2k) [inline models/error.go:1407]: I don't understand this. If the message doesn't start with "remote: ", the rest of the loop will never execute. Is that intended? What's a typical message here?

File: options/locale/locale_en-US.ini [hunk 3]
```diff
@@ -1086,6 +1086,7 @@ pulls.merge_conflict = Merge Failed: There was a conflict whilst merging: %[1]s<
 pulls.rebase_conflict = Merge Failed: There was a conflict whilst rebasing commit: %[1]s<br>%[2]s<br>%[3]s<br>Hint:Try a different strategy
 pulls.unrelated_histories = Merge Failed: The merge head and base do not share a common history. Hint: Try a different strategy
 pulls.merge_out_of_date = Merge Failed: Whilst generating the merge, the base was updated. Hint: Try again.
+pulls.push_rejected = Merge Failed: The push was rejected with the following message:<br>%s
```
Comment (bagasme) [inline options/locale/locale_en-US.ini]: Looks OK, but why adding `<br>`?

File: models/error.go (hunk 2 above)
Comment (zeripath) [reply models/error.go:1407]: We are deliberately only extracting the remote message here. The rest of the git bumpf is irrelevant.

File: models/error.go (hunk 1 above)
Comment (zeripath) [reply models/error.go]: That's a copy and paste error.

File: options/locale/locale_en-US.ini (hunk 3 above)
Comment (zeripath) [reply options/locale/locale_en-US.ini]: It forces the message on to a new line. Probably we should be putting these into a `<p>` but IIRC that breaks the rendering

File: models/error.go (hunk 2 above)
Comment (guillep2k) [reply models/error.go:1407]: I mean, shoudln't that `break` be `i++` and `continue` instead? The message I get doesn't even contain `"remote: "`:

```
Failed to update/create file 'vaafallar' with error: Unable to push back to repo from temporary repo: user1/publicrepo (/home/gittest/data/tmp/local-repo/upload.git026502734) Error: exit status 1 - To /home/gittest/gitea-repositories/user1/publicrepo.git
 ! [remote rejected] 3c35192d4bf4c5fb936309f4d56cd5c203d37297 -> master (pre-receive hook declined)
error: failed to push some refs to '/home/gittest/gitea-repositories/user1/publicrepo.git'
```

And from command line:
```
$ git push
Enumerating objects: 3, done.
Counting objects: 100% (3/3), done.
Delta compression using up to 2 threads
Compressing objects: 100% (2/2), done.
Writing objects: 100% (2/2), 222 bytes | 222.00 KiB/s, done.
Total 2 (delta 1), reused 0 (delta 0)
To git.mydomain.com:user1/publicrepo.git
 ! [remote rejected] master -> master (pre-receive hook declined)
error: failed to push some refs to 'gittest@git.mydomain.com:user1/publicrepo.git'
```

Comment (zeripath) [reply models/error.go:1407]: Then your git hook is just rejecting without returning a message contrary to good practice.

File: models/error.go (hunk 1 above)
Comment (zeripath) [reply models/error.go]: done

File: models/error.go (hunk 2 above)
Comment (zeripath) [reply models/error.go:1407]: I've adjusted the return.

Comment (guillep2k) [reply models/error.go:1407]: > 
> 
> Then your git hook is just rejecting without returning a message contrary to good practice.

Right. My bad. 😄

File: models/error.go [hunk 4]
```diff
@@ -1371,6 +1372,48 @@ func (err ErrMergePushOutOfDate) Error() string {
...
+		}
+		i = i + 8
+		nl := strings.IndexByte(err.StdErr[i:], '\n')
```
Comment (guillep2k) [inline models/error.go:1410]: I'm assuming there's no chance for EOL style problems here. 😅

**Review by guillep2k — APPROVED** [review]

File: models/error.go (hunk 4 above)
Comment (zeripath) [reply models/error.go:1410]: I think git handles that for us.

File: modules/repofiles/temp_repo.go [hunk 5]
```diff
@@ -242,8 +242,28 @@ func (t *TemporaryUploadRepository) CommitTreeWithDate(author, committer *models
...
+	if err := git.NewCommand("push", t.repo.RepoPath(), strings.TrimSpace(commitHash)+":refs/heads/"+strings.TrimSpace(branch)).RunInDirTimeoutEnvPipeline(env, -1, t.basePath, stdout, stderr); err != nil {
+		errString := stderr.String()
+		if strings.Contains(errString, "non-fast-forward") {
+			return models.ErrMergePushOutOfDate{
+				StdOut: stdout.String(),
+				StdErr: errString,
+				Err:    err,
+			}
+		} else if strings.Contains(errString, "! [remote rejected]") {
+			log.Error("Unable to push back to repo from temporary repo due to rejection: %s (%s)\nStdout: %s\nError: %v",
```
Comment (guillep2k) [inline modules/repofiles/temp_repo.go]: Maybe also log `errString` as well?

**Review by guillep2k — APPROVED** [review]
Just a nit, if you want...

File: modules/repofiles/temp_repo.go (hunk 5 above)
Comment (zeripath) [reply modules/repofiles/temp_repo.go]:
```suggestion
			log.Error("Unable to push back to repo from temporary repo due to rejection: %s (%s)\nStdout: %s\nStderr: %s\nError: %v",
```

File: modules/repofiles/temp_repo.go [hunk 6]
```diff
@@ -242,8 +242,28 @@ func (t *TemporaryUploadRepository) CommitTreeWithDate(author, committer *models
...
+				t.repo.FullName(), t.basePath, stdout, err)
```
Comment (zeripath) [inline modules/repofiles/temp_repo.go]:
```suggestion
				t.repo.FullName(), t.basePath, stdout, errString, err)
```

**Comment by zeripath** [comment]
Because this PR fixes the sanitization of the error message in the Web editor I've marked this as backport/v1.11

**Review by lafriks — APPROVED** [review]

**Comment by 6543** [comment]
pleace send backport

**Pull request merged**